From ff0f5284df93f21e509e406c14b524a200591835 Mon Sep 17 00:00:00 2001
From: Ang Tien Sung <tien.sung.ang@intel.com>
Date: Wed, 24 Nov 2021 08:42:50 +0800
Subject: [PATCH 17/20] HSD #15010184463: crypto: intel_fcs: Fix the inability
 to probe driver

commit d7a151522f5506ff8f6950037d79dc8cdc479f80 from
https://github.com/altera-opensource/linux-socfpga/commits/socfpga-5.4.124-lts

Properly define fcs device in the device tree and add proper driver
registeration and initialization for this change.

Signed-off-by: Ang Tien Sung <tien.sung.ang@intel.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 arch/arm64/boot/dts/intel/socfpga_agilex.dtsi |  4 ++
 drivers/crypto/intel_fcs.c                    | 43 ++++++++++++++++++-
 2 files changed, 46 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/intel/socfpga_agilex.dtsi b/arch/arm64/boot/dts/intel/socfpga_agilex.dtsi
index a9a9fb789d19..5d6b3ad189a2 100644
--- a/arch/arm64/boot/dts/intel/socfpga_agilex.dtsi
+++ b/arch/arm64/boot/dts/intel/socfpga_agilex.dtsi
@@ -630,6 +630,10 @@
 				fpga_mgr: fpga-mgr {
 					compatible = "intel,agilex-soc-fpga-mgr";
 				};
+
+				fcs: fcs {
+					compatible = "intel,agilex-soc-fcs";
+				};
 			};
 		};
 	};
diff --git a/drivers/crypto/intel_fcs.c b/drivers/crypto/intel_fcs.c
index 9dfab957b38c..1b193d9d966b 100644
--- a/drivers/crypto/intel_fcs.c
+++ b/drivers/crypto/intel_fcs.c
@@ -2486,15 +2486,56 @@ static int fcs_driver_remove(struct platform_device *pdev)
 	return 0;
 }
 
+static const struct of_device_id fcs_of_match[] = {
+	{.compatible = "intel,stratix10-soc-fcs"},
+	{.compatible = "intel,agilex-soc-fcs"},
+	{},
+};
+
 static struct platform_driver fcs_driver = {
 	.probe = fcs_driver_probe,
 	.remove = fcs_driver_remove,
 	.driver = {
 		.name = "intel-fcs",
+		.of_match_table = of_match_ptr(fcs_of_match),
 	},
 };
 
-module_platform_driver(fcs_driver);
+MODULE_DEVICE_TABLE(of, fcs_of_match);
+
+static int __init fcs_init(void)
+{
+	struct device_node *fw_np;
+	struct device_node *np;
+	int ret;
+
+	fw_np = of_find_node_by_name(NULL, "svc");
+	if (!fw_np)
+		return -ENODEV;
+
+	of_node_get(fw_np);
+	np = of_find_matching_node(fw_np, fcs_of_match);
+	if (!np) {
+		of_node_put(fw_np);
+		return -ENODEV;
+	}
+
+	of_node_put(np);
+	ret = of_platform_populate(fw_np, fcs_of_match, NULL, NULL);
+	of_node_put(fw_np);
+	if (ret)
+		return ret;
+
+	return platform_driver_register(&fcs_driver);
+}
+
+static void __exit fcs_exit(void)
+{
+	return platform_driver_unregister(&fcs_driver);
+}
+
+module_init(fcs_init);
+module_exit(fcs_exit);
 
 MODULE_LICENSE("GPL v2");
 MODULE_DESCRIPTION("Intel FGPA Crypto Services Driver");
-- 
2.31.1

